core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java

```diff
@@ -558,97 +558,138 @@ public CreatePathResult createPath(LockedInodePath inodePath, CreatePathOptions<
+          lockList.unlockLast();
+
+          dir =
+              (InodeDirectory) currentInodeDirectory.getChildReadLock(pathComponents[k], lockList);
+          if (dir == null) {
+            // Could not get the child inode. Continue and try again.
+            continue;
+          }
+        } else {
+          // Successfully added the child, while holding the write lock.
+          dir.setPinned(currentInodeDirectory.isPinned());
+          if (options.isPersisted()) {
+            // Do not journal the persist entry, since a creation entry will be journaled instead.
+            InodeUtils.syncPersistDirectory(dir, this, mMountTable, NoopJournalContext.INSTANCE);
+          }
+          // Journal the new inode.
+          journalContext.append(dir.toJournalEntry());
+          mInodes.add(dir);
+
+          // After creation and journaling, downgrade to a read lock.
+          lockList.unlockLast();
+
+          dir =
+              (InodeDirectory) currentInodeDirectory.getChildReadLock(pathComponents[k], lockList);
+          if (dir == null) {
+            // Could not get the child inode. Continue and try again.
+            continue;
+          }
+        }
       }
+
       createdInodes.add(dir);
-      mInodes.add(dir);
+      extensibleInodePath.getInodes().add(dir);
       currentInodeDirectory = dir;
     }
 
     // Create the final path component. First we need to make sure that there isn't already a file
     // here with that name. If there is an existing file that is a directory and we're creating a
     // directory, update persistence property of the directories if needed, otherwise, throw
     // FileAlreadyExistsException unless options.allowExists is true.
-    Inode<?> lastInode = currentInodeDirectory.getChild(name);
-    if (lastInode != null) {
-      // Lock the last inode before subsequent operations, and add it to the lock group.
-      lockList.lockWriteAndCheckNameAndParent(lastInode, currentInodeDirectory, name);
-
-      if (lastInode.isDirectory() && options instanceof CreateDirectoryOptions && !lastInode
-          .isPersisted() && options.isPersisted()) {
-        // The final path component already exists and is not persisted, so it should be added
-        // to the non-persisted Inodes of traversalResult.
-        existingNonPersisted.add(lastInode);
-        toPersistDirectories.add(lastInode);
-      } else if (!lastInode.isDirectory() || !(options instanceof CreateDirectoryOptions
-          && ((CreateDirectoryOptions) options).isAllowExists())) {
-        String errorMessage = ExceptionMessage.FILE_ALREADY_EXISTS.getMessage(path);
-        LOG.error(errorMessage);
-        throw new FileAlreadyExistsException(errorMessage);
+    Inode<?> lastInode = null;
+    while (lastInode == null) {
+      // Try to lock the last inode with the lock mode of the path.
+      if (extensibleInodePath.getLockMode() == LockMode.READ) {
```

Comment:
what do you think about using a switch statement here to make it explicit that we get write-lock for both `LockMode.WRITE` and `LockMode.WRITE_PARENT`?